src/ast/trapping.h

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_ast_trapping_h
+#define wasm_ast_trapping_h
+
+#include "pass.h"
+
+namespace wasm {
+
+enum class TrapMode {
+  Allow,
+  Clamp,
+  JS
+};
+
+inline void addTrapModePass(PassRunner& runner, TrapMode trapMode) {
+  if (trapMode == TrapMode::Clamp) {
+    runner.add("trap-mode-clamp");
+  } else if (trapMode == TrapMode::JS) {
+    runner.add("trap-mode-js");
+  }
+}
+
+class GeneratedTrappingFunctions {
```

Comment:
`TrappingFunctionGenerator` perhaps?

Reply:
Somewhat misleading because it doesn't do the generation directly. `TrappingFunctionContainer` maybe?

Reply:
`TrappingFunctionManager`?

Reply:
`XManager` sounds too general to me, I hear it saying "I do many X-related things, none are specific".
This class is responsible for making sure functions get generated only once, and adding them to the module at the appropriate time. It does not explicitly generate them, so `Builder` is out. `Container` isn't quite right but I'm not sure what's most precise.
`Storage`? ` Cache`? `Stash`? thesaurus.com isn't really helping me here.
Naming things is hard.